browser/models.py

```diff
@@ -3,83 +3,664 @@
 from __future__ import division
 from __future__ import print_function
 
+import copy
+import io
+import json
 import logging
-import pickle
+import tarfile
+import tempfile
 import timeit
 
+import boto3
 from flask_sqlalchemy import SQLAlchemy
+from flask import current_app
+from matplotlib import pyplot as plt
+import numpy as np
+from skimage.exposure import rescale_intensity
+from sqlalchemy.ext.compiler import compiles
+
+
+from helpers import is_npz_file, is_trk_file
+from config import AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
+from imgutils import pngify, add_outlines
 
 
 logger = logging.getLogger('models.Project')  # pylint: disable=C0103
 db = SQLAlchemy()  # pylint: disable=C0103
 
 
+@compiles(db.PickleType, "mysql")
+def compile_pickle_mysql(type_, compiler, **kw):
+    """
+    Replaces default BLOB with LONGBLOB for PickleType columns on MySQL backend.
+    BLOB (64 kB) truncates pickled objects, while LONGBLOB (4 GB) stores it in full.
+    TODO: change to MEDIUMBLOB (16 MB)?
+    """
+    return "LONGBLOB"
+
+
 class Project(db.Model):
     """Project table definition."""
     # pylint: disable=E1101
     __tablename__ = 'projects'
     id = db.Column(db.Integer, primary_key=True, autoincrement=True)
-    filename = db.Column(db.Text, nullable=False)
-    state = db.Column(db.LargeBinary(length=(2 ** 32) - 1))
-    subfolders = db.Column(db.Text, nullable=False)
     createdAt = db.Column(db.TIMESTAMP, nullable=False, default=db.func.now())
-    updatedAt = db.Column(db.TIMESTAMP, nullable=False, default=db.func.now(),
-                          onupdate=db.func.current_timestamp())
     finished = db.Column(db.TIMESTAMP)
-    numUpdates = db.Column(db.Integer, nullable=False, default=0)
-    firstUpdate = db.Column(db.TIMESTAMP)
-    lastUpdate = db.Column(db.TIMESTAMP)
 
-    def __init__(self, filename, state, subfolders):
-        self.filename = filename
-        self.state = state
-        self.subfolders = subfolders
+    raw_frames = db.relationship('RawFrame', backref='project')
+    rgb_frames = db.relationship('RGBFrame', backref='project')
+    label_frames = db.relationship('LabelFrame', backref='project')
+    # Includes underscore to avoid conflict with metadata (MetaData) attribute from SQLAlchemy
+    metadata_ = db.relationship('Metadata', backref='project', uselist=False)
+
+    def __init__(self, filename, input_bucket, output_bucket, path,
+                 rgb=False, raw_key=None, annotated_key=None):
+        init_start = timeit.default_timer()
+        if raw_key is None:
+            raw_key = 'raw'
```

Comment:
Why not just use as default `raw_key='raw'`?

Reply:
I think I just did the same thing for both raw and annotated, but it was only needed for annotated. Fixed to put raw_key='raw' in the function signature instead.